File: solutions/367.valid-perfect-square/valid-perfect-square.java
class Solution {
    public boolean isPerfectSquare(int num) {
        if(num < 0) return false;
        long start = 0, end = num;
        while(start <= end){
            long mid = (end - start)/2 + start;
            if(mid*mid == num){
                return true;
            }else if(mid*mid > num){
                end = mid - 1;
            }else{
                start = mid + 1;
            }
        }
        return false;
    }
}

File: solutions/367.valid-perfect-square/valid-perfect-square.js
/**
 * @param {number} num
 * @return {boolean}
 */
var isPerfectSquare = function(num) {
    let start = 1, end = num + 1;
    let res = false;
    while(start < end){
        let mid = parseInt(start + (end - start) / 2, 10);
        if(mid*mid === num){
            res = true;
            break;
        }else if(mid*mid > num){
            end = mid;
        }else{
            start = mid + 1;
        }
    }
    return res;
};